frame/staking/src/ledger.rs

```diff
@@ -0,0 +1,414 @@
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::{
+	traits::{Defensive, Get, LockableCurrency, WithdrawReasons},
+	BoundedVec, CloneNoBound, EqNoBound, PartialEqNoBound, RuntimeDebugNoBound,
+};
+use scale_info::TypeInfo;
+use sp_runtime::{traits::Zero, Perquintill, Rounding, Saturating};
+use sp_staking::{EraIndex, OnStakingUpdate, StakingAccount};
+use sp_std::{collections::btree_map::BTreeMap, prelude::*};
+
+use crate::{log, BalanceOf, Bonded, Config, Error, Ledger, UnlockChunk, STAKING_ID};
+
+/// The ledger of a (bonded) stash.
+///
+/// Note: All the reads and mutations to the `Ledger` and `Bonded` storage items *MUST* be performed
+/// through the methods exposed by this struct to ensure data and staking lock consistency.
+#[derive(
+	PartialEqNoBound,
+	EqNoBound,
+	CloneNoBound,
+	Encode,
+	Decode,
+	RuntimeDebugNoBound,
+	TypeInfo,
+	MaxEncodedLen,
+)]
+#[scale_info(skip_type_params(T))]
+pub struct StakingLedger<T: Config> {
+	/// The stash account whose balance is actually locked and at stake.
+	pub stash: T::AccountId,
+	/// The total amount of the stash's balance that we are currently accounting for.
+	/// It's just `active` plus all the `unlocking` balances.
+	#[codec(compact)]
+	pub total: BalanceOf<T>,
+	/// The total amount of the stash's balance that will be at stake in any forthcoming
+	/// rounds.
+	#[codec(compact)]
+	pub active: BalanceOf<T>,
+	/// Any balance that is becoming free, which may eventually be transferred out of the stash
+	/// (assuming it doesn't get slashed first). It is assumed that this will be treated as a first
+	/// in, first out queue where the new (higher value) eras get pushed on the back.
+	pub unlocking: BoundedVec<UnlockChunk<BalanceOf<T>>, T::MaxUnlockingChunks>,
+	/// List of eras for which the stakers behind a validator have claimed rewards. Only updated
+	/// for validators.
+	pub claimed_rewards: BoundedVec<EraIndex, T::HistoryDepth>,
+	/// The controller associated with this ledger's stash.
+	///
+	/// This is not stored on-chain, and is only bundled when the ledger is read from storage.
+	/// Use [`controller`] function to get the controller associated with the ledger.
+	#[codec(skip)]
+	controller: Option<T::AccountId>,
+}
+
+/// Represents the status of a ledger.
+pub(crate) enum StakingLedgerStatus<T: Config> {
```

Comment:
I recently learned about type state pattern in rust which you might find interesting to improve the `StakingLedger` api further.

https://www.youtube.com/watch?v=_ccDqRTx-JU

Reply:
Great suggestion. I think that the main advantage of using the type pattern here is that we can simplify the API by removing the results/options returned when the ledger is bonded. The new API would look something long the lines of:

```rust
pub struct Bonded;
pub struct Unbonded;

pub struct StakingLedger<T: Config, State = Unbonded> {
	// snip
	state: std::marker::PhantomData<State>,
}

impl <T: Config> StakingLedger<T, Unbonded> {
	pub(crate) fn bond(self) {}
}

impl <T: Config> StakingLedger<T, Boonded> {
	pub(crate) fn paired_account(account: StakingAccount<T::AccountId>) -> T::AccountId
	
	pub(crate) fn controller(&self) -> T::AccountId

	pub(crate) fn update(&self)

	pub(crate) fn kill(stash: &T::AccountId)

	pub(crate) fn consolidate_unlocked(self, current_era: EraIndex) -> Self

	pub(crate) fn rebond(mut self, value: BalanceOf<T>) -> (Self, BalanceOf<T>)
}

impl<T: Config> StakingLedger<T> {
	pub(crate) fn new() -> StakingLedger<T> {}

	// returnes a bonded account, if it exists.
	pub(crate) fn get(account: StakingAccount<T::AccountId>) -> Option<StakingLedger<T, Bonded>>

	// maybe not needed anymore
	pub(crate) fn is_bonded(account: StakingAccount<T::AccountId>) -> bool {}
}
```

I will give it a shot and update the PR if the improvements end up being worth it. Thanks!

Reply:
GIven that it's not trivial to define generics in the Ledger storage map due to the storage macros, I think that we should skip this idea for now.